Q1. What is the difference between a function and a method in Python ?
    
    - In Python, functions and methods are both blocks of reusable code that perform a specific task, but they differ mainly in how they are defined and used.

    Function -
    
    A function is defined using the def keyword and is not associated with an object (unless it's inside a class).



  Independent: Not tied to any object.
  Called directly: greet("Chetan").


In [122]:
# Example:

def greet(name):
  return f"Hello, {name}!"



In [123]:
print(greet("Chetan"))           # output : Hello, Chetan!

Hello, Chetan!


Method -

A method is a function that belongs to an object (usually a class instance).

1) Belongs to an object or class.
2) Called on an object: p.greet("Chetan")
3) The first parameter is typically self, referring to the instance itself.

In [124]:
# Example:


class Person:
  def greet(self, name):
    return f"Hello,{name}!"

p = Person()
print(p.greet("Chetan"))          # Output : Hello, Chetan!

Hello,Chetan!


Q2. Explain the concept of function arguments and parameters in Python.
    - In Python, function arguments and parameters are related but slightly different concepts used when defining and calling functions.

    Parameters

    Parameters are the placeholders in the function definition. They specify what kind of values the function expects to receive.

    

In [171]:
# Example:

def greet(name):
    print("Hello", name)


Here, name is a parameter.

Arguments

Arguments are the actual values you pass into a function when you call it.



In [172]:
# Example:

greet("Chetan")


Hello Chetan


Here, "Chetan" is an argument passed to the function.

Types of Arguments in Python:

1) Positional Arguments
Passed in the same order as parameters.

In [173]:
def add(x, y):
    return x + y
add(5, 3)  # 5 and 3 are positional arguments


8

2) Keyword Arguments
Passed using the name of the parameter.

In [174]:
add(x=5, y=3)


8

3) Default Arguments

Parameters with default values.

In [175]:
def greet(name="Guest"):
    print("Hello", name)
greet()           # Uses default value
greet("Chetan")   # Overrides default


Hello Guest
Hello Chetan


Variable-length Arguments

Accept multiple values.
*args for non-keyworded variable arguments (tuple).
**kwargs for keyworded variable arguments (dictionary).


In [176]:
def demo(*args, **kwargs):
    print(args)     # Tuple of arguments
    print(kwargs)   # Dictionary of keyword arguments

demo(1, 2, 3, a=4, b=5)


(1, 2, 3)
{'a': 4, 'b': 5}


Q3.  What are the different ways to define and call a function in Python ?

    - In Python, there are several ways to define and call functions, depending on the use case. Below are the main types and examples of each:

    1] Standard Function Definition

In [125]:
# Definition:

def greet(name):
  return f"Hello,{name}!"

In [126]:
# Calling:

print(greet("Alice"))

Hello,Alice!


2] Function with Default Arguments

In [127]:
# Definition:

def greet(name = "Guest"):
  return f"Hello, {name}!"

In [128]:
# Calling:

print(greet("Bob"))     # Uses  "Bob"
print(greet())

Hello, Bob!
Hello, Guest!


3] Function with Arbitrary Arguments.

In [129]:
# *a. args (Non-keyword arguments)

def add_numbers(*args):
  return sum(args)


In [130]:
 print(add_numbers(1, 2, 3, 4))        # Output : 10

10


In [131]:
# **b. kwargs (Keyword arguments)

def print_info(**kwargs):
  for key, value in kwargs.items():
    print(f"{key}:{value}")

In [132]:
print_info(name = "Alice", age = 25)

name:Alice
age:25


4] Lambda Function (Anonymous Function)


  Use for short, simple functions — commonly with map(), filter(), etc.

In [133]:
# Definition & Calling:

square = lambda x: x*x
print(square(5))            # Output : 25

25


5] Nested Functions (Function inside a Function)

In [134]:
# Example:


def outer():
  def inner():
    return "Hello from inner!"
  return inner()


In [135]:
print(outer())

Hello from inner!


6] Function as First-Class Object

You can assign a function to a variable, pass it as an argument, or return it.

In [136]:
# Example:


def shout(text):
  return text.upper()

def whisper(text):
  return text.lower()

def greet(func):
  return func("Hi There")




In [137]:
print(greet(shout))       # HI THERE
print(greet(whisper))     #hi there

HI THERE
hi there


7] Recursion (Function Calling Itself)

In [138]:
# Example:

def factorial(n):
  if n == 1:
    return 1
  return n * factorial(n - 1)



In [139]:
print(factorial(5))       # Output : 120

120


8] Calling Function from Another Module



In [140]:
# Example:
# Instead of importing, define it directly here


def say_hello():
  return "Hello!"

print(say_hello())




Hello!


Q4. What is the purpose of the `return` statement in a Python function ?

    - The return statement in a Python function serves the following main purposes:

    1]Sends a value back to the caller
    When a function is called, it can compute and return a value using return.

In [141]:
# Example:


def add(a, b):
  return a + b



In [142]:
result = add(5, 3)
print(result)           # Output : 8

8


2] Ends the function execution

As soon as a return statement is executed, the function exits immediately, even if more code follows.

In [143]:
def example():
  return "Done"

print("This will not run")          # This line is never executed

This will not run


3] Returns multiple values (as a tuple)

Python allows returning multiple values using a single return.

In [144]:
def operations(a, b):
  return a + b
  return a - b


In [145]:
sum_result

15

In [146]:
diff_result

5

4] Returns None by default

If no return is specified, the function returns None.

In [147]:
# Example:

def greet():
  print("Hello")

In [148]:
result = greet()
print(result)         # Output : None

Hello
None


Q5. What are iterators in Python and how do they differ from iterables ?

    - In Python, iterators and iterables are two closely related but distinct concepts that are central to looping and data traversal.

    Iterables-

    An iterable is any Python object capable of returning its members one at a time.
    Examples: list, tuple, string, set, dictionary, etc.
    1] An iterable implements the __iter__() method.
    2] You can use it in a for loop.
    3] It doesn't keep track of where it is in the iteration.

In [149]:
# Example:

my_list = [1, 2, 3]
for item in my_list:        # my_list is iterable
  print(item)


1
2
3


Iterators -

An iterator is an object that represents a stream of data; it returns data one element at a time when you call next() on it.

It implements both:

__iter__() (returns the iterator object itself),

__next__() (returns the next item or raises StopIteration).

It remembers its position during iteration.
It’s consumed as you iterate.

In [150]:
# Example:

my_iter = iter([1, 2, 3])  # Convert iterable to iterator
print(next(my_iter))  # 1
print(next(my_iter))  # 2
print(next(my_iter))  # 3
# print(next(my_iter))  # Raises StopIteration


1
2
3


In [151]:
# Example:
#Example to Illustrate


# Iterable
my_str = "abc"
print(iter(my_str))  # <str_iterator object>

# Iterator
it = iter(my_str)
print(next(it))  # 'a'
print(next(it))  # 'b'
print(next(it))  # 'c'


a
b
c


Q6.  Explain the concept of generators in Python and how they are defined ?
    
    - Generators are a special type of iterator in Python that allow you to iterate over data lazily—meaning they generate values on the fly and don't store everything in memory at once. This makes them memory-efficient for large datasets or infinite sequences.

    Key Characteristics:

    1) Lazy Evaluation: Values are computed only when needed.
    2) One-time Use: Generators can be iterated only once.
    3) Efficient: They don’t store all values in memory.

    How Generators are Defined:
    There are two main ways to create generators in Python:

    1] Using a Generator Function (with yield)

    yield pauses the function and saves its state.

    When next() is called again, it resumes from where it left off.
    

In [152]:
def count_up_to(n):
  count = 1
  while count <= n:
    yield count           # Suspend the function returns value
    count += 1

In [153]:
gen = count_up_to(3)
for num in gen:
  print(num)

1
2
3


2] Using Generator Expressions (similar to list comprehensions)

      Uses round brackets () instead of square brackets [].
      More compact but less flexible than generator functions.

      Under the Hood:

      Generators implement the iterator protocol (__iter__() and __next__()).
      They raise StopIteration when exhausted.

      When to Use Generators:
      Working with large data (like reading big files line by line).
      When performance and memory are concerns.
      For streaming data or infinite sequences (e.g., Fibonacci, prime numbers, etc.).
      

In [154]:
gen = (x**2 for x in range (5))
for num in gen:
  print(num)

0
1
4
9
16


Q7.  What are the advantages of using generators over regular functions ?

    - Generators offer several advantages over regular functions, especially when working with large data or when performance and memory efficiency are important. Here are the main benefits:

    1] Memory Efficiency

    Generators don’t store the entire sequence in memory. Instead, they yield items one at a time.
    Useful for working with large datasets or infinite sequences.


    This avoids creating a list with n elements.

In [155]:
# Example:


def count_up_to(n):
  for i in range(n):
    yield i

2] Lazy Evaluation

Values are generated on the fly, only when needed.
Saves time and memory during iteration.

In [156]:
# Example:

for num in count_up_to(1_000_000):
  if num == 10:
    break

Stops early without computing the rest.

3] Performance Boost

Faster startup time compared to regular functions that return lists.
Reduces overhead of memory allocation.

4] State Persistence
Generators automatically remember their state between yield s.
Useful for implementing pipelines, coroutines, or stateful sequences.

5] Simplifies Code for Iterators
Easier than writing an iterator class with __iter__() and __next__() methods.
Cleaner and more readable code for sequential processes.

6]  Composable
You can chain generators to build data pipelines (like Unix pipes).

In [157]:
# Example:

def even_numbers(seq):
  for num in seq:
    if num % 2 == 0:
      yield num

Q8. What is a lambda function in Python and when is it typically used ?

    - A lambda function in Python is an anonymous, inline function defined using the lambda keyword. It is used to create small, one-line functions without formally defining them using def.

    

In [158]:
# Syntax:

lambda arguments: expression


<function __main__.<lambda>(arguments)>

In [159]:
# Example:

add = lambda x, y: x + y
print(add(2, 3))            # Output : 5

5


Key Characteristics:

No def or function name required.
Can take any number of arguments.
Must contain a single expression (no statements).
Returns the value of that expression.


Common Use Cases:
1] With map():

In [160]:
# Example:

nums = [1, 2, 3, 4]
squared = list(map(lambda x: x**2, nums))
print(squared)                                      #[1, 4, 9, 16]

[1, 4, 9, 16]


2] With filter():

In [161]:
# Example:

nums = [1, 2, 3, 4]
evens = list(filter(lambda x: x % 2 == 0, nums))
print(evens)                                          #[2, 4]

[2, 4]


3] With sorted() and custom key:

In [162]:
# Example:

names = ['John', 'Alexandra', 'Bob']
sorted_names = sorted(names, key = lambda name: len(name))
print(sorted_names)                            # ['Bob', 'John', 'Alexandra']

['Bob', 'John', 'Alexandra']


4] Inline callbacks or simple function logic:

Useful where defining a full function is unnecessary or overkill.

When not to use lambda:
If the logic is complex or requires multiple statements.
When readability and debugging are a priority — in those cases, use def.

Q9.  Explain the purpose and usage of the `map()` function in Python ?

      - The map() function in Python is a built-in function used to apply a given function to every item in an iterable (like a list, tuple, etc.) and return a map object (an iterator) of the results.

      Purpose of map()
      To transform data by applying a function to each element of a collection without using an explicit for loop.

      # Syntax:
      map(function, iterable)

      function: A function to apply to each item in the iterable.
      iterable: A sequence (e.g., list, tuple) whose items will be processed.

 Using lambda with map()

 numbers = [1, 2, 3, 4, 5 ]
 squared = map(lambda x: x ** 2, numbers)
 print(list(squared)
 # Output : [1, 4, 9, 16, 25]

In [163]:
# Basic Example:

def square(x):
    return x * x

numbers = [1, 2, 3, 4, 5]
squared_numbers = map(square, numbers)

print(list(squared_numbers))
# Output: [1, 4, 9, 16, 25]


[1, 4, 9, 16, 25]


Using map() with Multiple Iterables



In [164]:
a = [1, 2, 3]
b = [4, 5, 6]

In [165]:
sum_ab = map(lambda x, y: x + y, a, b)
print(list(sum_ab))

# Output : [5, 7, 9]

[5, 7, 9]


Key Points:

Returns a map object, which is an iterator (convertible to list, tuple, etc.).
More efficient than using loops for large data transformations.
Works well with lambda, built-in functions, or custom functions.

Q10. What is the difference between `map()`, `reduce()`, and `filter()` functions in Python ?

  - In Python, the built-in functions map(), reduce(), and filter() are commonly used for functional programming. Here's a breakdown of their differences and use cases:

  1] map()
  Purpose: Applies a function to each item in an iterable and returns a map object (an iterator).

Syntax:
map(function, iterable)


In [166]:
# Example:

nums = [1, 2, 3, 4]
squared = map(lambda x: x ** 2, nums)
print(list(squared))        # Output : [1, 4, 9, 16]

[1, 4, 9, 16]


Use case: Transform elements one-by-one.

2] filter()

Purpose: Filters items in an iterable based on a condition provided by a function (which returns True or False).

Syntax: filter(function, iterable)


In [167]:
# Example:

nums = [1, 2, 3, 4, 5]
even = filter(lambda x: x % 2 == 0, nums)
print(list(even))       # Output : [2, 4]

[2, 4]


Use case: Select elements based on a condition.

3] reduce() (requires functools module)
Purpose: Reduces an iterable to a single value by applying a function cumulatively.

Syntax: from functools import reduce
reduce(function, iterable)


In [168]:
# Example:

from functools import reduce
nums = [1, 2, 3, 4]
product = reduce(lambda x, y: x * y, nums)
print(product)          # Output : 24

24


Use case: Combine all elements to a single result (like sum, product, etc.).

Q11. Using pen & Paper write the internal mechanism for sum operation using  reduce function on this given
list:[47,11,42,13];

    - To understand the internal mechanism of the reduce() function with a sum operation on the list [47, 11, 42, 13], let's break it down step-by-step using pen & paper style:

    Concept:

    The reduce() function from Python’s functools module applies a rolling computation to sequential pairs of elements in a list.

    Goal:

    Compute the sum of the list:[47, 11, 42, 13]
    using:

In [169]:
# Example:

from functools import reduce
reduce(lambda a, b: a + b, [47, 11, 42, 13])


113

Step-by-step Reduction:

Initial list:
[47, 11, 42, 13]

Step 1:
Apply function on first two elements:
a = 47, b = 11
→ 47 + 11 = 58

Intermediate result = 58

Step 2:
Now take the result and next item:
a = 58, b = 42
→ 58 + 42 = 100

Intermediate result = 100

Step 3:
Next:
a = 100, b = 13
→ 100 + 13 = 113

Final result = 113



[link text](https://drive.google.com/file/d/1Uc2_DVLXSr9UKtgSb3yj4TAmJH072si5/view?usp=drive_link)